docs/how-to-guides/effective-gno.md

```diff
@@ -0,0 +1,255 @@
+use either panic or errors, depending on what you need.
+
+```go
+import "std"
+
+func Foobar() {
+    caller := std.GetOrigCaller()
+    if caller != "g1234567890123456789012345678912345678" {
+        panic("permission denied")
+    }
+    // ...
+}
+```
+
+- TODO: suggest MustXXX and AssertXXX flows in p/.
+
+### Understand the importance of `init()`
+
+In Gno, the `init()` function isn't just a function, it's a cornerstone. It's
+automatically triggered when a new realm is added onchain, making it a one-time
+setup tool for the lifetime of a realm.
+
+Unlike Go, where `init()` is used for tasks like setting up database
+connections, configuring logging, or initializing global variables every time
+you start a program, in Gno, `init()` is executed once per realm's lifetime.
+
+In Gno, `init()` primarily serves two purposes:
+1. It registers your new realm on a new realm. This is typically done using the
+   registry pattern. This means you import another realm and call a method.
+2. It configures the initial state, i.e., global variables.
+
+```go
+import "gno.land/r/some/registry"
+
+func init() {
+    registry.Register("myID", myCallback)
+}
+
+func myCallback(a, b string) { /* ... */ }
+```
+
+A common use case could be to set the "admin" as the caller uploading the
+package.
+
+```go
+import (
+    "std"
+    "time"
+)
+
+var (
+    created time.Time
+    admin   std.Address
+    list    []string
+)
+
+func init() {
+    created = time.Now()
+    admin = std.GetOrigCaller()
+    list = append(list, "foo", "bar")
+}
+```
+
+In essence, `init()` in Gno is your go-to function for setting up and
+registering realms. It's a powerful tool that helps keep your realms organized
+and properly configured from the get-go.
+
+## Gno Good Practices
+
+### Design Your Realm as a Public API
+
+In Go, all your packages, including your dependencies, are typically treated as
+part of your safe zone, similar to a secure perimeter. The boundary is drawn
+between your program and the rest of the world, which means you secure the API
+itself, potentially with authentication middlewares.
+
+However, in Gno, your package is the public API. It's exposed to the outside
+world and can be accessed by other realms. Therefore, it's crucial to design
+your realm with the same level of care and security considerations as you would
+a public API.
+
+One approach is to simulate a secure perimeter within your realm by having
+private functions for the logic and then writing your API layer by adding some
+front-facing API with authentication. This way, you can control access to your
+realm's functionality and ensure that only authorized callers can execute
+certain operations.
+
+```go
+import "std"
+
+func PublicMethod(nb int) {
+    caller := std.GetOrigCaller()
+    privateMethod(caller, nb)
+}
+
+func privateMethod(caller std.Address, nb int) { /* ... */ }
+```
+
+In this example, `PublicMethod` is a public function that can be called by other
```

Comment:
This is an awesome example, but the explanatory code block which _could_ showcase a modifier like `assertIsAdmin()` is missing (or hidden) within the function above `func privateMethod(caller std.Address, nb int) { /* ... */ }`. 

Perhaps we could expand the internal function's logic to showcase how the `caller` argument is checked against a custom modifier function that provides the necessary access control.

I think providing something here would make this example much more impactful, otherwise we are _assuming_ that developers will know what to do with `caller`; if they do nothing to check against `caller`, then calling `privateMethod` inside of `PublicMethod` will do little to prevent abuse given the missing access control logic.

Reply:
The pattern I proposed is actually implemented in some instances, for example, in foo20: https://github.com/gnolang/gno/blob/master/examples/gno.land/r/demo/foo20/foo20.gno#L50.

Your suggestion is valid, but I plan to address that aspect later. The primary intention behind this effective Gno advice is to maintain the internal API's privacy by simply making it unexported. This approach centers on a limited number of public functions. These functions could serve as straightforward proxies or incorporate additional logic. This might include adding a caller, a block number, or as you mentioned, conducting ACL checks.

I prefer to retain the example as it is for now. However, I agree that expanding the comments to include more examples could be beneficial if you think it would aid in understanding.